salt/modules/restartcheck.py

```diff
@@ -314,23 +314,40 @@ def _kernel_versions_nilrt():
             List with possible names of last installed kernel
             as they are probably interpreted in output of `uname -a` command.
     '''
-    kernel_versions = []
-
-    if __grains__.get('os_family') == NILRT_FAMILY_NAME and _is_older_nilrt():
-        # bzImage is copied in the rootfs without any package management or
-        # version info. We also can't depend on kernel headers like
-        # include/generated/uapi/linux/version.h being installed. Even if
-        # we fix this in newer versions of "old NILRT" we still need to be
-        # backwards compatible so it'll just get more complicated.
-        kpath = '/boot/runmode/bzImage'
-        kernel_strings = __salt__['cmd.run']('strings {0}'.format(kpath))
-        re_result = re.search(r'[0-9]+\.[0-9]+\.[0-9]+-rt.*(?=\s\()', kernel_strings)
-        if re_result is not None:
-            kernel_versions.append(re_result.group(0))
+    kver = None
+
+    def _get_kver_from_bin(kbin):
+        '''Get kernel version from a binary image or None if detection fails'''
+        kvregex = r'[0-9]+\.[0-9]+\.[0-9]+-rt.*(?=\s\()'
+        kernel_strings = __salt__['cmd.run']('strings {0}'.format(kbin))
+        re_result = re.search(kvregex, kernel_strings)
+        return None if re_result is None else re_result.group(0)
+
+    if _is_older_nilrt():
```

Comment:
Oh, very nice! `if distribution_i_use_at_my_office():` anyone? :wink:

@10ne1 since seems NILRT does not have a proper versioning, then can we come up with something generic instead, like property checker etc, which will do decision based on that (say, "foo is installed", "bar is located there" etc) instead of hardcoding "older version" or "even_older_version" right into the code?

Reply:
I think I can modify the distro rootfs without breaking backwards compatibility such that the lsb_release DISTRIB_ID field reflects the difference between these older/newer versions and that would automatically get picked up by the existing lsb_distrib_id grain, so then we'd just have to test for grains[lsb_distrib_id] == 'nilrt'.

Are you ok with this approach?

Reply:
@isbm A collegue of mine posted a PR implementing my suggestion, if that gets merged we can test the lsb grain in this one https://github.com/saltstack/salt/pull/47951